### PR TITLE
[PS-97]: Extend RaygunEnvironmentMessage to add Maui specific information.

### PR DESCRIPTION
**Pull request by Hamish-taylor**
## The problem
The problem this pull request is trying to solve is that the Raygun4net environment message is limited and does not contain mobile specific fields that other Raygun mobile specific providers support. We would like to support these fields in the Maui provider. This pr also covers populating the fields with the appropriate data.

## The implementation
There are two areas that were changed to enable this support, firstly a new `RaygunMauiEnvironmentMessage` class was created that extends the raygun4net.core `RaygunEnvironmentMessage` class. This new class contains the extra fields mentioned above. Secondly a `RaygunClientTest` (this name is only temporary) class was created extending `RaygunClient`. This class overrides the `BuildMessage` method, adding custom code to build the message with the new `RaygunMauiEnvironmentMessage` class.  

**Note**
This PR proposes a approach to building the `RaygunMessage` that breaks from convention. Typically builder classes have been used to create and populate the message and its child classes. This is how it is done inside the Raygun4net.core provider. 
My argument against this approach is that this creates unnecessary abstractions away from the data. Many of the builder methods just taking in a field as a parameter and store it inside of the underlying data structure, I propose eliminating the middle layer and instead to work with the data its self. In my approach the objects are explicitly created and populated, this makes it very clear what is going on and makes it very easy to modify in the future. With the builder approach, if you want to change one of the underlying datatypes, for example the `RaygunEnvironmentMessage`, you then have to change the `RaygunEnvironmentMessageBuilder` and the `RaygunMessageBuilder`. With this approach you simply have to change the `EnvironmentMessage` object you are creating and populate any extra fields. Given how small the code is to explicitly construct the objects without the builder I do not believe this is a concern that needs to be split of into its own classes. KISS

Author to check:

- [x] Builds pass
- [ ] Tested in an alternative environment
- [x] Reviewed by another developer
- [x] Appropriate documentation written
- [x] Tested on Andoid
- [x] Tested on iOS
- [x] Tested on MacOS
- [x] Tested on Windows

Reviewer to check:
 - [ ] Change has been tested on Beta
 - [ ] Code is written to standards
 - [ ] Appropriate tests have been written

### CODE REVIEW

File: Raygun4Maui/NativeDeviceInfo.cs [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+            return RuntimeInformation.ProcessArchitecture.ToString();
+#elif MACCATALYST
+             return RuntimeInformation.ProcessArchitecture.ToString();
+#else
+            return RuntimeInformation.ProcessArchitecture.ToString();
+#endif
+        }
+#if IOS || MACCATALYST
+        //Code from https://github.com/dannycabrera/Get-iOS-Model
+        public static string GetModel(string hardware)
```
Comment (QuantumNightmare) [inline Raygun4Maui/NativeDeviceInfo.cs:194]: I would highly recommend that the provider passes raw data to Raygun, and then this kind of conversion occurs on Raygun's side when processing or displaying the data as needed. That way, our customers won't need to keep updating the Maui provider and releasing new versions of their apps to support picking up on what devices their end users are using.
I've found that storing raw data and then converting it when displaying has been a good rule of thumb. (Sometimes you also store converted values to help with searching/filtering functionality).

File: README.md [hunk 2]
```diff
@@ -76,17 +76,17 @@ builder
 
 Unhandled exceptions will be sent to Raygun automatically.
 
-Raygun4Maui stores an instance of a [Raygun4Net](https://github.com/MindscapeHQ/raygun4net/tree/master/Mindscape.Raygun4Net.NetCore). `RaygunClient` object, this can be accessed through the following code:
+Raygun4Maui stores an instance of a `RaygunMauiClient` object that is initialised by the Maui builder, this can be accessed through the following code:
```
Comment (PanosNB) [inline README.md:79]: Not a big deal but we must be using American, not British, English for docs. So, 'initialized' :-)

File: Raygun4Maui/RaygunMauiClient.cs [hunk 3]
```diff
@@ -24,17 +28,70 @@ internal static void Attach(RaygunClient client)
...
+
+        public RaygunMauiClient(RaygunSettings settings) : base(settings)
+        {
+        }
+
+        public static readonly RaygunClientMessage clientMessage = new()
+        {
+            Name = "Raygun4Maui",
+            Version = Assembly.GetExecutingAssembly().GetName().Version.ToString(),
+            ClientUrl = "https://github.com/MindscapeHQ/raygun4maui"
```
Comment (PanosNB) [inline Raygun4Maui/RaygunMauiClient.cs:45]: Ideally, we don't want hardcoding of parameters, especially inside methods.  The Version could stay here to ensure it gets updated everytime it's called. You can either roll back to
```
private static readonly string clientName = "Raygun4Maui";
private static readonly string clientUrl = "https://github.com/MindscapeHQ/raygun4maui";
```
O you can put these in a settings file the app reads from

File: Raygun4Maui/RaygunMauiClient.cs [hunk 4]
```diff
@@ -24,17 +28,70 @@ internal static void Attach(RaygunClient client)
...
+        };
 
-        private static void OnSendingMessage(object sender, RaygunSendingMessageEventArgs e)
+        protected override async Task<RaygunMessage> BuildMessage(Exception exception, IList<string> tags, IDictionary userCustomData, RaygunIdentifierMessage userInfo)
```
Comment (PanosNB) [inline Raygun4Maui/RaygunMauiClient.cs:48]: Ad discussed, wrap in try-catch blocks for information that might throw exceptions

**Review by PanosNB — CHANGES_REQUESTED** [review]
As discussed, also update the readme with platform-specific interpretations of total memory

File: README.md [hunk 5]
```diff
@@ -131,7 +131,17 @@ try {
 
 ---
 ## Platform specific information
-Raygun4Maui will automatically collect information specific to the environment the application is being run in. However, on iOS, Raygun4Maui cannot obtain the devices name. This is a privacy restriction put in place by Apple. If you would like this information to be collected and sent with crash reports you will have to [request for permission from apple](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_device-information_user-assigned-device-name?language=objc).
+Raygun4Maui will automatically collect information specific to the environment the application is being run in. However, there are inconsistency's between certain values across platforms.
```
Comment (PanosNB) [inline README.md:134]: use plural: "inconsistencies"

File: README.md [hunk 6]
```diff
@@ -131,7 +131,17 @@ try {
...
+- on iOS, Raygun4Maui cannot obtain the devices name. This is a privacy restriction put in place by Apple. If you would like this information to be collected and sent with crash reports you will have to [request for permission from apple](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_device-information_user-assigned-device-name?language=objc).
```
Comment (PanosNB) [inline README.md:135]: "device's name"

**Review by PanosNB — APPROVED** [review]
Great work!, just a couple of typos in the readme

**Pull request merged**